Feat / Liquidity mining strategy

Nullably opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
https://www.notion.so/hummingbot/Liquidity-Mining-strategy-0c6eb43ac5ba4c4685e3092f404e0492

mifeng commented:
LGTM - I suggest adding a Markets section to `status` command that shows the base/quote balances and allocation %, like:

| Market | Base Balance | Quote Balance | Mid Price | Base/Quote % 
|-------|---------------|----------------|------|--------------|
AVAX-USDT | 0 | 190 | 0.78 | 0% / 100% |
HARD-USDT | 10 | 190 | 12.23 | 5% / 95% |

<img width="443" alt="Screen Shot 2021-01-25 at 6 52 57 PM" src="https://user-images.githubusercontent.com/1795030/105793875-c5701480-5f3e-11eb-8dc2-fb935608fc7a.png">

Nullably commented:
> LGTM - I suggest adding a Markets section to `status` command that shows the base/quote balances and allocation %, like:
> 
> Exchange	Market	Mid Price	Base Balance	Quote Balance	Base/Quote %
> binance	AVAX-USDT	0.78	0	190	0% / 100%
> binance	HARD-USDT	12.23	10	190	5% / 95%
> <img alt="Screen Shot 2021-01-25 at 6 52 57 PM" width="443" src="https://user-images.githubusercontent.com/1795030/105793875-c5701480-5f3e-11eb-8dc2-fb935608fc7a.png">

@mifeng a question:
From your example, shouldn't the Base/Quote % for HARD-USDT be 39% / 61% ? As we need to factor in the price as well, so the Base% is (12.23 * 10) / (190 + (12.23 * 10))

mifeng commented:
Oh my example was just for show, so I didn't calculate it correctly. Yes,
that makes sense.

On Thu, Jan 28, 2021, 9:54 PM Nullably <notifications@github.com> wrote:

> LGTM - I suggest adding a Markets section to status command that shows
> the base/quote balances and allocation %, like:
>
> Exchange Market Mid Price Base Balance Quote Balance Base/Quote %
> binance AVAX-USDT 0.78 0 190 0% / 100%
> binance HARD-USDT 12.23 10 190 5% / 95%
> [image: Screen Shot 2021-01-25 at 6 52 57 PM]
> <https://user-images.githubusercontent.com/1795030/105793875-c5701480-5f3e-11eb-8dc2-fb935608fc7a.png>
>
> @mifeng <https://github.com/mifeng> a question:
> From your example, shouldn't the Base/Quote % for HARD-USDT be 39% / 61% ?
> As we need to factor in the price as well, so the Base% is (12.23 * 10) /
> (190 + (12.23 * 10))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CoinAlpha/hummingbot/pull/2871#issuecomment-769593752>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANWHVUUAJFNJX4AG6W6BF3S4JEQ7ANCNFSM4WRXOTKA>
> .
>

rxlxrxsx commented:
After producing two trades on SCRT-ETH, this suddenly appeared on the right log pane, producing non-existing trades for QKC-ETH.
<img width="1422" alt="Screen Shot 2021-02-01 at 7 54 41 PM" src="https://user-images.githubusercontent.com/64377055/106456062-ff559380-64c7-11eb-9d4d-3856a079cc01.png">
<img width="434" alt="Screen Shot 2021-02-01 at 7 56 12 PM" src="https://user-images.githubusercontent.com/64377055/106456101-09779200-64c8-11eb-8933-8f61b812c350.png">
<img width="442" alt="Screen Shot 2021-02-01 at 7 56 23 PM" src="https://user-images.githubusercontent.com/64377055/106456109-0c728280-64c8-11eb-934e-7eae877d3ab5.png">

Trade history from Binance
<img width="1185" alt="Screen Shot 2021-02-01 at 7 49 36 PM" src="https://user-images.githubusercontent.com/64377055/106456117-0e3c4600-64c8-11eb-945f-e2add0f0310f.png">

.csv file shows trade data for QKC-ETH instead of trades that happened on SCRT-ETH
<img width="1095" alt="Screen Shot 2021-02-01 at 8 06 15 PM" src="https://user-images.githubusercontent.com/64377055/106456944-337d8400-64c9-11eb-8733-6f5b2fcc72b3.png">

Logs: [logs_lmstrat.log](https://github.com/CoinAlpha/hummingbot/files/5903808/logs_lmstrat.log)
Config: [lmstrat.txt](https://github.com/CoinAlpha/hummingbot/files/5903811/lmstrat.txt)
Trades: [trades_lmstrat.csv.zip](https://github.com/CoinAlpha/hummingbot/files/5903812/trades_lmstrat.csv.zip)

Nullably commented:
@keithbaum 
Can you please help take a look into this issue that @rxlxrxsx faced.
> After producing two trades on SCRT-ETH, this suddenly appeared on the right log pane, producing non-existing trades for QKC-ETH.

mifeng commented:
<img width="644" alt="Screen Shot 2021-02-01 at 4 21 25 PM" src="https://user-images.githubusercontent.com/1795030/106534731-93643280-64a9-11eb-91cc-ec127b0563fc.png">

- Use `Base Bal` instead of `Base Balance`, same for Quote
- Only display `Base %`
- Quote balance should take into account the allocation to that market - here, it's using my entire BNB balance of 4.99 instead of splitting it between the 2 markets
- Use the word `Amount` instead of `Size` since parameter is called `order_amount`
- Replace `Size ($)` with `Size (quote symbol)` since user inputs it in quote terms

Finally, if the bot adjusts spreads due to volatility, I think there should be a notification in `status` and a log message. We'll also need to document how this works, since it's not clear from the parameter names how users should use it.

Nullably commented:
@rxlxrxsx 
> Also, no notification is sent on the left pane when non partial orders are filled
I added the notification

@mifeng 
I have updated the branch to include your change requests.
> Replace Size ($) with Size (quote symbol) since user inputs it in quote terms
Since the token can also be a base token, hence there are many quote symbols. In this case, I just display `Size (Quote)`.

I added `Volatility` to `markets`, hope this will make it clearer. 
I will update the notion more on how volatility is calculated and how it is applied.

Nullably commented:
@rxlxrxsx 
> After producing two trades on SCRT-ETH, this suddenly appeared on the right log pane, producing non-existing trades for QKC-ETH

The fix for this is in this PR, so we'll have to wait for it to merge to development, then I'll include the change into this branch.
https://github.com/CoinAlpha/hummingbot/pull/2910

rxlxrxsx commented:
Strategy is using `order_size` parameter name instead of `order_amount`

<img width="403" alt="Screen Shot 2021-02-03 at 4 27 02 PM" src="https://user-images.githubusercontent.com/64377055/106722537-0d7aef80-6641-11eb-9fe5-88e2d0da7e8d.png">

Nullably commented:
@rxlxrxsx 
> Strategy is using order_size parameter name instead of order_amount
Fixed!

JeremyKono commented:
A suggestion that there should be a message in the logs saying if you don't have enough balance after running `start` command. As to avoid troubleshooting with support and this may cause confusion as to why the user's strategy is not running.

keithbaum commented:
PR #2910 is merged. Bug should be fixed now

CrimsonJacket reviewed (approved):
Ran the strategy overnight and confirmed that volatility adjustments were performed accordingly. The default `spread` was set to 0.4%
![image](https://user-images.githubusercontent.com/6755922/106823747-70f63300-66bc-11eb-8e5c-2ec8dc7d292f.png)

LGTM!

rxlxrxsx commented:
> PR #2910 is merged. Bug should be fixed now

Still getting the same behavior on the latest commit. There's also one filled order that got logged twice upon checking.

<img width="696" alt="Screen Shot 2021-02-04 at 4 15 56 PM" src="https://user-images.githubusercontent.com/64377055/106869286-b6415180-670a-11eb-9f52-acbecaaef527.png">
<img width="1136" alt="Screen Shot 2021-02-04 at 4 17 22 PM" src="https://user-images.githubusercontent.com/64377055/106869300-bd685f80-670a-11eb-9353-a7e9342f0df6.png">

Based on the logs the duplicate order showed after this message appeared.
```
2021-02-04 15:28:59,985 - 4385 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1612423739.0, "order_id": "x-XEKWYICX-SQCEH1612423712002325", "trading_pair": "QKC-ETH", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT_MAKER", "price": "0.00000635", "amount": "3012.00000000", "trade_fee": ["0", [["BNB", "0.00045153"]]], "exchange_trade_id": 2419936, "leverage": 1, "position": "NILL", "event_name": "OrderFilledEvent", "event_source": "binance"}
```
```
2021-02-04 15:29:00,490 - 4385 - hummingbot.connector.exchange.binance.binance_exchange - INFO - The market order x-XEKWYICX-SQCEH1612423712002325 was already filled, or order was not submitted by hummingbot.Ignoring trade filled event in update_order_status.
```

Logs: [logs_liquidity_mining.log](https://github.com/CoinAlpha/hummingbot/files/5924512/logs_liquidity_mining.log)
Config: [liquidity_mining.txt](https://github.com/CoinAlpha/hummingbot/files/5924515/liquidity_mining.txt)
csv: [trades_liquidity_mining.csv.zip](https://github.com/CoinAlpha/hummingbot/files/5924520/trades_liquidity_mining.csv.zip)
sqlite: [liquidity_mining.sqlite.zip](https://github.com/CoinAlpha/hummingbot/files/5924522/liquidity_mining.sqlite.zip)

```
commit: adc01739e242ae6ad59c95ea0aec1f638ace496e
Date: Thu Feb 4 07:42:38 2021 +0800
```

Nullably commented:
@keithbaum Can you please help look into the latest reported issue above?
Thanks.

keithbaum commented:
Hi @rxlxrxsx, about your remarks:
* Thanks for your testing. For the mixed trading pairs bug, I found the cause and address it in another PR #2924 .
* In regards to the trades you see duplicated, they are actually different amounts which sum up to the order (just like in the binance trade history). This order filling is not like the regular in_flight_orders since it's being pulled out from the exchange history, but numbers add up.
* You should be able to test in a branch with code from my PR merged in liquidity mining branch. Think it would easier if my PR could be merged with development branch and then merged into this PR for testing (Unfortunately the only way I see to test if issue from PR #2924 was fixed is using the liquidity mining strategy)

Nullably commented:
@keithbaum Feel free to make changes to this branch if that makes it easier for you to work on.

PtrckM commented:
- All issue above are fix. Tested and confirmed working by QA

![image](https://user-images.githubusercontent.com/54516858/106996102-3b843f00-67bb-11eb-8fd5-741812ab3dce.png)
![image](https://user-images.githubusercontent.com/54516858/106996119-44751080-67bb-11eb-9844-ad9596039007.png)

CrimsonJacket reviewed (approved):
LGTM! :tea:

mifeng commented:
@Nullably I think `Base Bal %` here is dividing by the total balance, not the amount allocated to the individual market. In the screenshot below, my ALGO-USDT base allocation should be 8%, not 2%.
<img width="526" alt="Screen Shot 2021-02-06 at 2 25 32 PM" src="https://user-images.githubusercontent.com/1795030/107130982-94b9a480-6887-11eb-90fe-1246f29d1ffb.png">


This is a minor issue so it shouldn't block merging to `development`. We can fix it in the next release if necessary.

mifeng commented:
Another change I'd like to propose is that rather than spreading the selected asset (USDT, etc) evenly across all markets, the strategy should compute how much quote to allocate to each market, so that the total value (base + quote) is equal is each market. 

This is needed to prevent the bot from changing the allocations whenever the user stops and restarts the strategy.

Nullably commented:
> @Nullably I think `Base Bal %` here is dividing by the total balance, not the amount allocated to the individual market. In the screenshot below, my ALGO-USDT base allocation should be 8%, not 2%.

@mifeng The Quote Bal showed there is actually the balance allocated to each market, not actual balance in your account.
This is my USDT balance
![image](https://user-images.githubusercontent.com/37262506/107172465-63b0af80-6a00-11eb-8ce4-c3128222b62f.png)
And my markets
![image](https://user-images.githubusercontent.com/37262506/107172500-7a570680-6a00-11eb-957a-4e4e84fd7b32.png)
The Base % is round up to 0 decimal point, in your ALGO-USDT it should be around 1.68%

Nullably commented:
> Another change I'd like to propose is that rather than spreading the selected asset (USDT, etc) evenly across all markets, the strategy should compute how much quote to allocate to each market, so that the total value (base + quote) is equal is each market.
> 
> This is needed to prevent the bot from changing the allocations whenever the user stops and restarts the strategy.

Ok, good idea. Just to confirm the allocation is calculated as
1. Find total portfolio value
  1.1 If `token` is quote, this is (base1 * price) + (base2 * price) + .. + quote
  1.2 If `token` is base, this is (quote1 / price) + (quote2 / price) + .. + base
2. Each market is allocated with total_value / no of markets
3. For each market, add up allocation to the desired portion
  3.1 If `token` is quote, if (base x price) is lower than the portion, add `quote` to the amount desired
  3.2 If `token` is base, if (quote / price) is lower than the portion, add `base` to the amount desired

mifeng commented:
Yes, I think that is correct.

On Sun, Feb 7, 2021 at 7:55 PM Nullably <notifications@github.com> wrote:

> Another change I'd like to propose is that rather than spreading the
> selected asset (USDT, etc) evenly across all markets, the strategy should
> compute how much quote to allocate to each market, so that the total value
> (base + quote) is equal is each market.
>
> This is needed to prevent the bot from changing the allocations whenever
> the user stops and restarts the strategy.
>
> Ok, good idea. Just to confirm the allocation is calculated as
>
>    1. Find total portfolio value
>    1.1 If token is quote, this is (base1 * price) + (base2 * price) + ..
>    + quote
>    1.2 If token is base, this is (quote1 / price) + (quote2 / price) + ..
>    + base
>    2. Each market is allocated with total_value / no of markets
>    3. For each market, add up allocation to the desired portion
>    3.1 If token is quote, if (base x price) is lower than the portion,
>    add quote to the amount desired
>    3.2 If token is base, if (quote / price) is lower than the portion,
>    add base to the amount desired
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CoinAlpha/hummingbot/pull/2871#issuecomment-774848242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANWHVXY7HMXCZXNZN244CTS55OD7ANCNFSM4WRXOTKA>
> .
>


-- 
Michael Feng
mifeng@gmail.com
+1 908 938 8733
